File: 17.demo_python_tf/src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
# 手眼坐标转换
import rclpy
import math
# 导入 math 库用于提供角度和弧度转换函数
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster
# 作用: 静态变换广播器，用于发布两个坐标系之间的固定变换关系
# 用途: 当坐标系之间的相对位置不会改变时使用，比如机器人底座到激光雷达的固定偏移
from geometry_msgs.msg import TransformStamped
# 作用: 定义坐标变换的消息类型
# ros2 interface show geometry_msgs/msg/TransformStamped
# 这个消息描述了在stamp时刻，从frame_id坐标系到child_frame_id坐标系的3D变换关系。变换包含了平移和旋转两个部分。
# 与Transform的区别：
#     Transform: 只包含变换数据（平移+旋转）
#     TransformStamped: 包含变换数据 + 时间戳 + 坐标系信息

# 欧拉角转四元数的函数（替代tf_transformations）
def euler_to_quaternion(roll, pitch, yaw):
    """
    将欧拉角转换为四元数
    参数: roll, pitch, yaw (弧度)
    返回: [x, y, z, w] 四元数
    """
    cy = math.cos(yaw * 0.5)
    sy = math.sin(yaw * 0.5)
    cp = math.cos(pitch * 0.5)
    sp = math.sin(pitch * 0.5)
    cr = math.cos(roll * 0.5)
    sr = math.sin(roll * 0.5)

    w = cr * cp * cy + sr * sp * sy
    x = sr * cp * cy - cr * sp * sy
    y = cr * sp * cy + sr * cp * sy
    z = cr * cp * sy - sr * sp * cy

    return [x, y, z, w]


class StaticTFBroadcaster(Node):
# StaticTFBroadcaster(Node): 表示StaticTFBroadcaster类继承自Node类
# Node是父类（超类/基类）
# StaticTFBroadcaster是子类（派生类）
# super().__init__('static_tf2_broadcaster'):
# 调用父类Node的__init__方法
# 传递节点名称'static_tf2_broadcaster'
    def __init__(self):
        super().__init__('static_tf2_broadcaster')
        # super(): Python的内置函数，用于访问父类
            # 返回一个代理对象，允许你调用父类的方法
            # 在多重继承中特别有用
        # __init__(): 构造函数/初始化方法
            # 每个Python类都有的特殊方法
            # 创建对象实例时自动调用
        # 相当于访问Node类的__init__，static_tf2_broadcaster为节点名

        self.static_broadcaster_ = StaticTransformBroadcaster(self)
        # StaticTransformBroadcaster(self): 创建静态变换广播器对象
        # StaticTransformBroadcaster: 从tf2_ros导入的类
        # (self): 将当前节点对象作为参数传递

        self.publish_static_tf()
         
    def publish_static_tf(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "base_link"
        transform.child_frame_id = "camera_link"
        transform.transform.translation.x = 0.5
        transform.transform.translation.y = 0.3  # 修正：这里应该是y坐标
        transform.transform.translation.z = 0.6  # 修正：这里应该是z坐标

        # 欧拉角转化为四元数（使用自定义函数）
        rotation_quart = euler_to_quaternion(math.radians(180), 0, 0)

        transform.transform.rotation.x = rotation_quart[0]
        transform.transform.rotation.y = rotation_quart[1]
        transform.transform.rotation.z = rotation_quart[2]
        transform.transform.rotation.w = rotation_quart[3]

        # 发布静态坐标转换
        # 转换并且发布
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f' 发布 TF:{transform}')



def main():
    rclpy.init()
    static_tf_broadcaster = StaticTFBroadcaster()  # 修正变量名
    rclpy.spin(static_tf_broadcaster)  # 修正变量名
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: 18.demo_cpp_tf/src/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include <memory>
#include "geometry_msgs/msg/transform_stamped.hpp" //提供消息接口
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"             //提供 tf2::Quaternion类
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp" //提供消息类型转换函数
#include "tf2_ros/static_transform_broadcaster.h"  //提供静态坐标广播器类

class StaticTFBroadcaster : public rclcpp::Node
{
public:
    StaticTFBroadcaster() : Node("tf_broadcaster_node")
    {
        // 创建静态广播发布器并发布
        // this在类的成员函数 / 构造函数里，是一个普通指针，指向当前类的实例对象；
        // 你的这个语句写在StaticTFBroadcaster的构造函数里，
        // 所以这里的this → 指向当前正在创建的 ROS2 节点对象（就是你后面main里std::make_shared<StaticTFBroadcaster>()创建的那个节点）；
        // 为什么要传this给广播器？
        // TF 广播器不是独立工作的，它需要依赖 ROS2 节点的核心资源（比如节点的时钟、通信上下文、命名空间等）才能正常发布 TF 坐标变换消息，
        // 简单说：告诉 TF 广播器 “你属于哪个节点，要跟着这个节点干活”，这是 ROS2 所有组件（发布器、订阅器、TF 广播器 / 监听器）的通用规则 —— 创建时必须绑定所属节点。
        broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        this->publish_tf();
    }

    void publish_tf()
    {
        geometry_msgs::msg::TransformStamped transform;
        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "target_point";
        transform.transform.translation.x = 5.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;
        tf2::Quaternion quat;
        quat.setRPY(0, 0, 60 * M_PI / 180); //弧度制欧拉角转四元数
        transform.transform.rotation = tf2::toMsg(quat);    //转成消息接口类型
        broadcaster_->sendTransform(transform);
    }

private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> broadcaster_;
};

int main(int argc,char** argv){
    rclcpp::init(argc,argv);
    auto node = std::make_shared<StaticTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

/*
使用ros2 run demo_cpp_tf static_tf_broadcaster 执行文件

$ ros2 run tf2_ros tf2_echo map target_point
---
At time 0.0
- Translation: [5.000, 3.000, 0.000]
- Rotation: in Quaternion [0.000, 0.000, 0.500, 0.866]
- Rotation: in RPY （radian） [0.000, -0.000, 1.047]
- Rotation: in RPY （degree） [0.000, -0.000, 60.000]
- Matrix:
0.500 -0.866 0.000 5.000
0.866 0.500 0.000 3.000
0.000 0.000 1.000 0.000
0.000 0.000 0.000 1.000
这个命令
ros2 run <功能包名> <可执行文件/节点名> [参数1] [参数2] ...
tf2_ros：功能包名，是 ROS2 处理 ** 坐标系变换（tf2）** 的核心功能包，提供了 tf2 相关的工具、节点和底层 API；
tf2_echo：该功能包下的可执行调试工具，核心功能是「回显 / 打印」两个坐标系之间的 tf 变换数据，且会实时刷新（坐标系变换更新时，输出也会同步更新）；
*/